src/sources/socket/mod.rs

```diff
@@ -0,0 +1,255 @@
+mod tcp;
+mod udp;
+
+use super::util::TcpSource;
+use crate::{
+    event::{self, Event},
+    topology::config::{DataType, GlobalOptions, SourceConfig, SourceDescription},
+};
+use futures::sync::mpsc;
+use serde::{Deserialize, Serialize};
+use std::net::SocketAddr;
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+// TODO: add back when serde-rs/serde#1358 is addressed
+// #[serde(deny_unknown_fields)]
+pub struct SocketConfig {
+    #[serde(flatten)]
+    pub mode: Mode,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(tag = "mode", rename_all = "snake_case")]
+pub enum Mode {
+    Tcp(tcp::TcpConfig),
+    Udp(udp::UdpConfig),
+}
+
+impl SocketConfig {
+    pub fn make_tcp_config(addr: SocketAddr) -> Self {
+        Self {
+            mode: Mode::Tcp(tcp::TcpConfig::new(addr.into())),
```

Comment:
With the implementation of `From` below, this could be just `mode: tcp::TcpConfig::new(addr.into()).into()`

Reply:
Done!